rclpy/rclpy/__init__.py

```diff
@@ -67,11 +67,12 @@ def spin_once(node):
 
     _rclpy.rclpy_wait(wait_set)
 
-    # TODO(wjwwood): properly implement this by checking the contents of the wait_set.
-    for subscription in node.subscriptions:
-        msg = _rclpy.rclpy_take(subscription.subscription_handle, subscription.msg_type)
+    sub_ready_list = _rclpy.rclpy_get_ready_subscriptions(wait_set)
+    for sub_idx in sub_ready_list:
+        sub_ready = node.subscriptions[sub_idx]
```

Comment:
From a multi threading standpoint this looks dangerous. What if `node.subscriptions` changes while the `rclpy_wait` function is blocking? I think something more sophisticated is needed to corrollate a subscription in the waitset from one in the node (other than a list index). Maybe you could use the underlying rcl_subscription_t pointer or the Python object's `id()` or a GUID of some sort.